File: app/src/main/java/com/ternaryop/photoshelf/birthday/BirthdayManager.java
package com.ternaryop.photoshelf.birthday;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;

import com.ternaryop.utils.JSONUtils;
import org.json.JSONArray;
import org.json.JSONObject;

/**
 * Created by dave on 01/04/17.
 * Image Extractor Manager
 */

public class BirthdayManager {
    public static final String API_PREFIX = "http://visualdiffer.com/image";

    private String accessToken;

    public BirthdayManager(String accessToken) {
        this.accessToken = accessToken;
    }

    public BirthdayInfo search(final String name) throws Exception {
        StringBuilder sb = new StringBuilder(API_PREFIX + "/v1/birthday/search?name=" + URLEncoder.encode(name, "UTF-8"));

        HttpURLConnection conn = null;
        try {
            conn = getSignedGetConnection(sb.toString());
            handleError(conn);
            final JSONArray birthdayInfo = toJson(conn.getInputStream()).getJSONArray("birthdays");
            return new BirthdayInfo(birthdayInfo.getJSONObject(0));
        } finally {
            if (conn != null) try { conn.disconnect(); } catch (Exception ignored) {}
        }
    }

    protected HttpURLConnection getSignedGetConnection(String url) throws IOException {
        HttpURLConnection conn = (HttpURLConnection)new URL(url).openConnection();
        conn.setRequestProperty("PhotoShelf-Subscription-Key", accessToken);
        conn.setRequestMethod("GET");

        return conn;
    }

    public JSONObject toJson(InputStream is) throws Exception {
        try (InputStream bis = new BufferedInputStream(is)) {
            return JSONUtils.jsonFromInputStream(bis);
        }
    }

    protected void handleError(HttpURLConnection conn) throws Exception {
        if (conn.getResponseCode() == 200) {
            return;
        }
        final JSONObject error = toJson(conn.getErrorStream());
        throw new RuntimeException("Error " + conn.getResponseCode() + ": " + error.getString("errorMessage"));
    }
}

File: app/src/main/java/com/ternaryop/photoshelf/birthday/BirthdayManager.kt
package com.ternaryop.photoshelf.birthday

import com.ternaryop.utils.JSONUtils
import org.json.JSONObject
import java.io.BufferedInputStream
import java.io.IOException
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

/**
 * Created by dave on 01/04/17.
 * Image Extractor Manager
 */

private const val API_PREFIX = "http://visualdiffer.com/image"

class BirthdayManager(private val accessToken: String) {

    @Throws(Exception::class)
    fun search(name: String): BirthdayInfo {
        val sb = "$API_PREFIX/v1/birthday/search?name=${URLEncoder.encode(name, "UTF-8")}"

        var conn: HttpURLConnection? = null
        return try {
            conn = getSignedGetConnection(sb)
            handleError(conn)
            val birthdayInfo = toJson(conn.inputStream).getJSONArray("birthdays")
            BirthdayInfo(birthdayInfo.getJSONObject(0))
        } finally {
            try {
                conn?.disconnect()
            } catch (ignored: Exception) {
            }
        }
    }

    @Throws(IOException::class) private fun getSignedGetConnection(url: String): HttpURLConnection {
        val conn = URL(url).openConnection() as HttpURLConnection
        conn.setRequestProperty("PhotoShelf-Subscription-Key", accessToken)
        conn.requestMethod = "GET"

        return conn
    }

    @Throws(Exception::class)
    private fun toJson(stream: InputStream): JSONObject {
        BufferedInputStream(stream).use { bis -> return JSONUtils.jsonFromInputStream(bis) }
    }

    @Throws(Exception::class) private fun handleError(conn: HttpURLConnection) {
        if (conn.responseCode == HttpURLConnection.HTTP_OK) {
            return
        }
        val error = toJson(conn.errorStream)
        throw RuntimeException("Error ${conn.responseCode} : ${error.getString("errorMessage")}")
    }
}
